[WIP] Pull LoopBack 4 doc updates from loopback-next

b-admike opened the pull request:
Connect to https://github.com/strongloop/loopback-next/issues/843. This script copies over the doc files for LoopBack 4 content into this repo, and reverts some linking changes made in `loopback-next` in order to render it properly when Jekyll is run. Mad credits to @rmg for essentially writing the script! The script is intended to demonstrate how to achieve the proposal in the spike, so it would probably need to be modified if we are to use this approach.

**EDIT** The script now utilizes `@loopback/docs` to pull in the latest changes from `loopback-next` for LoopBack 4 docs and `jekyll-relative-links` plugin to handle `.md` links (with one exception). It also provides a Travis config to deploy to GitHub Pages whenever there are new changes.

rmg commented:
Your `update-lb4-docs.sh` needs a shebang at the top (`#!/bin/bash`) to ensure that it is executed with the correct interpreter instead of relying on `/bin/sh` being the default.

rmg commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,19 @@
+   | tar --strip-components 1 -C $tmp -xzf- '*.md'
+cp -R $tmp/packages $target/readmes/$repo
+#rename relative link extensions from md to html
+sed -E -i 's/\(([^(]+)\.md/(\1.html)/g' $tmp/docs/*.md
```
You have:
```
s/\(([^(]+)\.md/(\1.html)/g
```
It should be this:
```
s/\(([^(]+)\.md\)/(\1.html)/g
```
You need to match the closing `)` literal otherwise it doesn't get replaced by the `)` literal in the replacement and you end up with double `))` on the output - and also incorrectly matching:
 `(index.md other stuff)` -> `(index.html) other stuff)`

b-admike replied:
Good catch. Actually I wanted to include the cases where `.md` is followed by `#` and left out the literal match for `)`. I will remove it from the replacement as well.

bschrammIBM reviewed (approved):
lgtm

Commit:
Retrieve doc updates from loopback-next

Add a script to copy over loopback-next docs using
@loopback/docs. Add Travis config to pick up the
changes and deploy to GitHub Pages upon a successful
build. Run README update scripts from Travis as well
and remove Jenkinsfile which did the same thing with
Jenkins.

shimks commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,17 @@
+curl -L https://github.com/$ORG/$REPO/archive/$COMMIT.tar.gz | tar --strip-components 1 -C "$DOCS" --wildcards \
+-xzf- "$ARCHIVE_NAME/docs/*.md" "$ARCHIVE_NAME/packages/*.md" "$ARCHIVE_NAME/docs/*.png"
+
+# cp -R $DOCS/packages $DOCS/readmes/$REPO
```
Was this meant to be commented out?

virkt25 replied:
+1

shimks reviewed (approved):
LGTM, but one comment

virkt25 commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,17 @@
+ORG=strongloop
+REPO=loopback-next
+DOCS=./pages/en/lb4
+COMMIT=893d68547766fc41d8bca82425645416e357d052
```
Is this meant to be hard-coded?

b-admike replied:
Good catch; it's supposed to be master.

raymondfeng commented on Jenkinsfile:
```diff
@@ -9,13 +9,18 @@ def gitEnv = [
   }
   stage('download updates') {
     sh './update-readmes.sh'
-    sh './update-v4-readmes.sh'
```
We probably should keep `sh './update-v4-readmes.sh'` here as it's used to pull in README files. Even after we move docs to the monorepo, we might continue to pull in README files.

raymondfeng commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,16 @@
+mkdir -p $DOCS
+# pull down all the markdown files (reserving dir structure)
+# given the GitHub ORG, REPO, and COMMIT
+curl -L https://github.com/$ORG/$REPO/archive/$COMMIT.tar.gz | tar --strip-components 1 -C "$DOCS" --wildcards \
```
At some point, it's probably better to switch to `npm install`

virkt25 commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,20 @@
+# -xzvf- "$ARCHIVE_NAME/docs/*.md" "$ARCHIVE_NAME/packages/*.md"
+# "$ARCHIVE_NAME/docs/*.png"
+npm i @loopback/docs --prefix $DOCS
+mv $DOCS/node_modules/@loopback/docs $DOCS/monorepo-docs
```
Is `/monorepo-docs` needed? Or can we just move the docs fo `$DOCS`?

raymondfeng replied:
The `site/*` should be copied into `pages/en/lb4/`.

b-admike replied:
I was just using it as a placeholder, before we implement changes that override the current doc contents, but yeah it should be `site/*`.

raymondfeng commented:
I just published `@loopback/docs@0.1.2`, which has all related files under `site` directory.

raymondfeng commented on update-lb4-docs.sh:
```diff
@@ -0,0 +1,20 @@
+#curl -L https://github.com/$ORG/$REPO/archive/$BRANCH.tar.gz | tar --strip-components 1 -C "$DOCS" --wildcards \
+# -xzvf- "$ARCHIVE_NAME/docs/*.md" "$ARCHIVE_NAME/packages/*.md"
+# "$ARCHIVE_NAME/docs/*.png"
+npm i @loopback/docs --prefix $DOCS
```
I think it can be further simplified if we add `@loopback/docs` as a dependency in `package.json` of `loopback.io`. This way, `npm install` at the root level will pull in `node_modules/@loopback/docs` and we can add a post-install script to copy them to the right place.
_

b-admike replied:
I talked to @rmg and if we aim to do that, then we should probably introduce Travis to this repo since the Jenkins pipeline is
>running directly on the Jenkins master and that server doesn't have any external executors 

and doesn't have `node/npm` installed. It would be a hassle to install them there. Also,
>all GitHub Pages does is run Jekyll and upload the generated `_site` directory (or whatever it is..) to S3

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,34 @@
+const destImages = path.resolve(__dirname, 'images', 'lb4');
+
+function copyDocs(src, dest) {
+    try {
```
IIRC, we should use two spaces for indentation.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,34 @@
+
+function copyDocs(src, dest) {
+    try {
+        fs.readdirSync(src).forEach((val, index) => {
```
I suggest that we use `fs-extra` to make the code much simpler.

virkt25 commented on package.json:
```diff
@@ -16,6 +18,7 @@
     "url": "git+https://github.com/Strongloop/loopback.io.git"
   },
   "devDependencies": {
+    "@loopback/docs": "^0.1.2",
```
Should we use `latest` or `*`?

shimks commented:
The `Edit this page` button in `_layouts/home.html` and `_layouts/page.html` need to be changed so that when the button is pressed, it links towards the pages in the monorepo and not the one in loopback.io

shimks commented:
I also wanted to know whether we want to bring over sidebar and some other miscellaneous stuff (one I could think off the top of my head are tables in app/artifact-generator.md). It makes sense to bring this over to the monorepo since these types of files need to be modified/changed as docs are being written/modified.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,42 @@
+  '@loopback',
+  'docs',
+  'site',
+);
```
It can be simplified as  `const srcDocs = path.resolve(__dirname, 'node_modules/@loopback/docs/site');`.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,42 @@
+
+async function copyDocs(src, dest) {
+  try {
+    await fs.copy(src, dest);
```
I think `fs.copySync` is good enough.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,42 @@
+fs
+  .ensureDir(path.resolve(destDocs, 'imgs'))
+  .then(() => {
+    fs.removeSync(path.resolve(destDocs, 'imgs'));
```
No need to check existence of `imgs` when we use `fs.removeSync` (rm -rf). See https://github.com/jprichardson/node-fs-extra/blob/master/docs/remove-sync.md.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,42 @@
+  fs.writeFileSync(fileToUpdate, contents, 'utf-8');
+} catch (err) {
+  console.error('failed to replace relative link %s', err.stack);
+}
```
If there is an error, we should make sure the process exit with a non-zero code so that `npm` fails.

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,37 @@
+
+const srcDocs = path.resolve(__dirname,'node_modules/@loopback/docs/site');
+const srcImages = path.resolve(srcDocs, 'imgs');
+const destDocs = path.resolve(__dirname, 'pages', 'en', 'lb4');
```
Same here, `pages/en/lb4`

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,37 @@
+const srcDocs = path.resolve(__dirname,'node_modules/@loopback/docs/site');
+const srcImages = path.resolve(srcDocs, 'imgs');
+const destDocs = path.resolve(__dirname, 'pages', 'en', 'lb4');
+const destImages = path.resolve(__dirname, 'images', 'lb4');
```
`images/lb4`

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,37 @@
+
+function copyDocs(src, dest) {
+  try {
+    await fs.copySync(src, dest);
```
`await` should be removed. It cannot be used within a regular function.

b-admike commented:
@slnode test please

raymondfeng commented on Jenkinsfile:
```diff
@@ -9,13 +9,19 @@ def gitEnv = [
   stage('download updates') {
     sh './update-readmes.sh'
     sh './update-v4-readmes.sh'
+    sh './update-lb4-docs.sh'
```
I only see `update-lb4-docs.js`

raymondfeng reviewed (approved):
Please squash the commits before merge.

raymondfeng commented:
@b-admike Do we want to delete existing files from `pages/en/lb4` and add this folder to `.gitignore`?

b-admike commented:
@raymondfeng I was thinking we overwrite them, but that is a good idea.

virkt25 commented on .travis.yml:
```diff
@@ -0,0 +1,39 @@
+    - stage: copy v2,v3 readmes
+      script: /bin/bash ./update-readmes.sh
+      after_success: skip
+      os:
```
In a matrix config, if the setting is the same as the one outside it doesn't need to be re-listed. So in other words, you can remove the `os` property from under each `stage`.

virkt25 commented on .travis.yml:
```diff
@@ -0,0 +1,39 @@
+  github-token: $GITHUB_TOKEN
+  keep-history: true
+  on:
+    branch: master
```
Needs newline

virkt25 reviewed (changes requested):
I don't think the Travis config will do what you're trying to accomplish. Each Job runs in it's own VM and as such the copied artifacts aren't shared across each job or stage (deploy is considered a stage). The scripts for moving the readme's should be run in the `deploy` stage. 

Reference: 
https://docs.travis-ci.com/user/build-stages/#Data-persistence-between-stages-and-jobs

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,37 @@
+const srcDocs = path.resolve(__dirname,'node_modules/@loopback/docs/site');
+const srcImages = path.resolve(srcDocs, 'imgs');
+const destDocs = path.resolve(__dirname, 'pages/en/lb4');
+const destImages = path.resolve(__dirname, 'images/lb4');
```
Now we copy markdown and image files to two different places. Is it going to be  a problem for image links such as `./imgs/my-diagram.png`?

b-admike replied:
That's a good point. I've removed `images/lb4` entirely, and the images will live in `pages/en/lb4/imgs`. That way we don't have to worry about image links too (I've verified it).

raymondfeng commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,37 @@
+
+function copyDocs(src, dest) {
+  try {
+    fs.copySync(src, dest);
```
I think we should call `fs.removeSync(dest);` first now.

shimks commented on update-lb4-docs.js:
```diff
@@ -0,0 +1,40 @@
+// copy the latest docs from @loopback/docs to pages/en/lb4 directory
+copyDocs(srcDocs, destDocs);
+
+const fileToUpdate = path.resolve(destDocs, 'Testing-the-API.md');
```
Obsolete code?

b-admike replied:
Nah, it is used down the line (see https://github.com/strongloop/loopback.io/pull/643/commits/1ba3aba52978255543499c0ceaa8f429e2f30427#diff-015a67e879f2083136bcd7796abb4f33R34)

bajtos commented on Jenkinsfile:
```diff
@@ -9,6 +9,11 @@ def gitEnv = [
 env.CHANGE_BRANCH = env.CHANGE_BRANCH ?: env.BRANCH_NAME
 
+properties([
+    // Run every hour
```
Running Jenkins job every hour, isn't that excessive use of resources? Ideally, we should use GitHub's HTTP hooks to trigger the run of this job right after loopback-next's master branch was changed, or perhaps use some sort of npm mechanism to run the job after a new version of `@loopback/docs` was published.

If there isn't any easy solution for triggering the run only when there are changes to pull, then I am proposing to change this frequence to something like once a day, or at least every 4 hours or so.

cc @rmg

b-admike replied:
Yeah I had a discussion with @rmg and running every hour isn't excessive from what he told me (correct me if I'm wrong, @rmg). But I would like us to use Travis instead of Jenkins for updating the docs, so this is probably outdated (I'll probably revert/disable this once we have Travis up and running).

bajtos commented on .travis.yml:
```diff
@@ -0,0 +1,32 @@
+    branch: pull-monorepo-docs #for testing purposes
+  include:
+    - stage: copy v2,v3 readmes
+      script: /bin/bash ./update-readmes.sh
```
I find it a bit confusing that we have both Travis CI and Jenkins configured for this repository. What is the interaction between these two CI servers? Could you please describe this setup in more details, e.g. in [README.md](https://github.com/strongloop/loopback.io/blob/gh-pages/README.md) or perhaps in a new file?

cc @rmg

b-admike replied:
Good point. I will take a crack at documenting it.

rmg commented on _layouts/page.html:
```diff
@@ -8,10 +8,14 @@
+   {% if site.lb4_editme_path and page.layout != 'readme' and page.sidebar == 'lb4_sidebar' %}
+       {% assign edit_url =  page.url | remove: "doc/en/lb4" | remove: ".html" | append: ".md" | prepend: site.lb4_editme_path %}
+       <a target="_blank" href="https://github.com/{{edit_url}}" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit this page</a>
+  {% endif %}
```
wrong indentation?

rmg commented on Jenkinsfile:
```diff
@@ -9,6 +9,7 @@ def gitEnv = [
 
 env.CHANGE_BRANCH = env.CHANGE_BRANCH ?: env.BRANCH_NAME
 
+
```
Looks like a random whitespace change was accidentally left in.

virkt25 commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+    - npm test
+  email: slnode@ca.ibm.com
+  name: StrongLoop Bot
+    
```
Think there's some extra empty space here.

rmg commented:
Travis has been configured

b-admike commented:
Unfortunately, we can't use custom commit messages for the commits made by Travis on successful builds, but the feature might come soon https://github.com/travis-ci/travis-ci/issues/9287.
